lib/Dialect/FIRRTL/FIRRTLFieldSource.cpp

```diff
@@ -26,119 +26,74 @@
-  sv.push_back(sf.getFieldIndex());
-  makeNodeForValue(sf.getResult(), node->src, sv, foldFlow(sf));
-}
-
-void FieldSource::visitSubindex(SubindexOp si) {
-  auto value = si.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(si.getIndex());
-  makeNodeForValue(si.getResult(), node->src, sv, foldFlow(si));
-}
-
-void FieldSource::visitOpenSubindex(OpenSubindexOp si) {
-  auto value = si.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(si.getIndex());
-  makeNodeForValue(si.getResult(), node->src, sv, foldFlow(si));
-}
-
-void FieldSource::visitSubaccess(SubaccessOp sa) {
-  auto value = sa.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(-1);
-  makeNodeForValue(sa.getResult(), node->src, sv, foldFlow(sa));
-}
-
-void FieldSource::visitMem(MemOp mem) {
-  for (auto r : mem.getResults())
-    makeNodeForValue(r, r, {}, foldFlow(r));
-}
-
-void FieldSource::visitInst(InstanceOp inst) {
-  for (auto r : inst.getResults())
-    makeNodeForValue(r, r, {}, foldFlow(r));
-}
-
-void FieldSource::visitInstChoice(InstanceChoiceOp inst) {
-  for (auto r : inst.getResults())
-    makeNodeForValue(r, r, {}, foldFlow(r));
-}
-
-const FieldSource::PathNode *FieldSource::nodeForValue(Value v) const {
+  // Check cache.
   auto ii = paths.find(v);
-  if (ii == paths.end())
-    return nullptr;
-  return &ii->second;
+  if (ii != paths.end())
+    return ii->second;
+  
+  Operation* op = v.getDefiningOp();
+  return TypeSwitch<Operation*, PathNode>(op)
+  .Case<SubfieldOp>([&](auto subf) {
+    auto parent = visitValue(subf.getInput(), update);
+    PathNode retval{parent.src, parent.fieldID + subf.getInput().getType().base().getFieldID(subf.getFieldIndex())};
+    update(v, retval);
+    return retval;
+  })
+  .template Case<OpenSubfieldOp>([&](auto subf) {
+    auto parent = visitValue(subf.getInput(), update);
+    PathNode retval{parent.src, parent.fieldID + subf.getInput().getType().getFieldID(subf.getFieldIndex())};
+    update(v, retval);
+    return retval;
+  })
+  .template Case<SubindexOp>([&](auto subi) {
+    auto parent = visitValue(subi.getInput(), update);
+    PathNode retval{parent.src, parent.fieldID + subi.getInput().getType().base().getFieldID(subi.getIndex())};
+    update(v, retval);
+    return retval;
+  })
+  .template Case<OpenSubindexOp>([&](auto subi) {
+    auto parent = visitValue(subi.getInput(), update);
+    PathNode retval{parent.src, parent.fieldID + subi.getInput().getType().getFieldID(subi.getIndex())};
+    update(v, retval);
+    return retval;
+  })
+  .template Case<WireOp, RegOp, RegResetOp, InstanceOp, InvalidValueOp, chirrtl::MemoryPortOp>([&](auto def) {
+    return PathNode{v, 0};
+  })
+  .Default([&](auto* op) {
```

Comment:
This looks a fair bit like FieldRefCache -- what's the difference? Not hard to imagine differences but at a glance this is a cache of root value + fieldID for values?

cc https://github.com/llvm/circt/pull/6181 .

Reply:
Or rather, since this predated it, FieldRefCache is a compy of this?

Reply:
Well it may be functionally a copy now, but FieldSource until this PR (so, presently) was about computing paths for these things and wasn't obviously related to FieldRef's (and I think did more useful things re:subaccess that made it more pointer-analysis-like?).  I'm not 100% they're the same now, hence the question, but now that the result tracked for each value looks FieldRef-like it's possible there is overlap.

But yes FieldSource predates FieldRefCache, FWIW/as you say.  Until this PR reworked it I hadn't thought of them as solving related problems, to be honest.

FieldRefCache is for replacing/consolidating the pattern of `DenseMap<Value,FieldRef> refs; refs[v] = getFieldRefFromValue(v)` (optionally looks through casts since I found that useful for some use case), and caching computed FieldRef's along the way to avoid ever computing a FieldRef more than once so as to avoid some possible worst-case behaviors resulting from querying long chains of indexing at different points.

Honestly the most important part of that PR IMO isn't the cache but rather refactoring things so we can re-use code doing this sort of thing (getDeltaRef, getAccessedField).

Mechanically, FieldRefCache is not currently an analysis and computes its per-value information on-demand by walking up use-def chains, which might have different behavior compared to walking the IR forward as done here.  If you only expect to query a small subset of the IR that's a trade-off.

Anyway, let's not have multiple versions of the same thing without making it clear what their differences are.

Especially if it IS functionally different in what it computes as the analysis result per-value (`PathNode != getFieldRefFromValue(v)`) that seems great to be clear on :+1: .